platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/services/DefaultIsolatedProjectEvaluationListenerProvider.kt

```diff
@@ -73,48 +84,96 @@ class DefaultIsolatedProjectEvaluationListenerProvider(
         } ?: action
 
     override fun isolateFor(gradle: Gradle): ProjectEvaluationListener? = when {
-        beforeProject.isEmpty() && afterProject.isEmpty() -> null
+        serializedIsolatedActions != null -> serializedIsolatedActions
+            ?.let { IsolatedProjectEvaluationListener(gradle, it) }
+            ?: isolateFor(gradle)
+
+        beforeProject.isEmpty() && afterProject.isEmpty() && allprojects.isEmpty() -> null
         else -> {
-            val isolate = isolate(
-                IsolatedProjectActions(beforeProject, afterProject),
-                IsolateOwners.OwnerGradle(gradle)
-            )
-            clear()
-            IsolatedProjectEvaluationListener(gradle, isolate)
+            val isolated = isolateActions(gradle)
+            serializedIsolatedActions = isolated
+            IsolatedProjectEvaluationListener(gradle, isolated)
+        }
+    }
+
+    override fun isolateAllprojectsActionFor(gradle: Gradle): IsolatedAction<in Project>? = when {
+        allprojects.isEmpty() -> null
+        else -> {
+            val isolated = serializedIsolatedActions ?: isolateActions(gradle)
+            Allprojects(gradle, isolated)
         }
     }
 
     override fun clear() {
         beforeProject.clear()
         afterProject.clear()
+        allprojects.clear()
+        serializedIsolatedActions = null
     }
 
     private
-    fun isolate(actions: IsolatedProjectActions, owner: IsolateOwner) =
+    fun isolateActions(owner: Gradle): SerializedIsolatedActionGraph<IsolatedProjectActions> =
+        isolate(
+            IsolatedProjectActions(allprojects, beforeProject, afterProject),
+            IsolateOwners.OwnerGradle(owner)
+        )
+
+    private
+    fun <T : Any> isolate(what: T, owner: IsolateOwner) =
         IsolatedActionSerializer(owner, owner.serviceOf(), owner.serviceOf())
-            .serialize(actions)
+            .serialize(what)
 }
 
 
 private
 data class IsolatedProjectActions(
+    val allprojects: IsolatedProjectActionList,
     val beforeProject: IsolatedProjectActionList,
     val afterProject: IsolatedProjectActionList
 )
 
 
+private
+class Allprojects(
+    private val gradle: Gradle,
+    private val isolated: SerializedIsolatedActionGraph<IsolatedProjectActions>
+) : IsolatedAction<Project> {
+
+    override fun execute(target: Project) {
+        // Check if allprojects already been applied
+        if (target.peekSingletonProperty<IsolatedProjectActions>() != null) return
```

Comment:
❓ Here the implementation hole:
`target` is coming here is `ProblemReportingProblem`, what is not allowing me to use `target.extensions` IP safely, because the action may happening in the context of referrer project, who has called `allprojects {}`.

It's possible not to pass `IsolatedProjectActions` through `Project` properties and manage this state "at side" tho, but I'm considering the current solution as elegant(do you?), so I'd explore possibilities to keep it, rather than invent internal state management for project status.

So your input is appriciated

Reply:
An example of the state management I'm talking about
https://github.com/gradle/gradle/pull/29565